README.md
# API2go

[![Build Status](https://travis-ci.org/quagliato/api2go.svg?branch=master)](https://travis-ci.org/quagliato/api2go)

## Description

Simple structure to create APIs based on HTTP requests over express.

You can create RESTful or RPC APIs, using any kind of HTTP method accepted by
expressjs and customize the paths you use.

### Why use API2Go instead of express?

API2Go abstracts a lot of things that in express you would have to implement
manually, like this:

- **Logger**: API2GO() has a **logger** function publicly accessible;
- **Email**: every instance of API2Go() has also a function **sendMail** 
publicly accessible;
- **Auditing**: the API2Go logs every request, the input, the output, how many
time it tooks to process etc.

### Some things that need to be implemented

Like any one-man-army, I can't do everything at the same time, so these are some
of the things that I plan to program or want to program:

- Middleware;
- Unified and strong authentication (can't continue using the payload);
- Headers filtering;
- N-level validation on functions' map;
- Got any idea? *eduardo@quagliato.me* or add a new issue in the Github repo;

## Usage

    npm install api2go

* * * * *

To use (**from v0.0.13 and on**):
    
    var API2Go = require('api2go');

    // You can instance N API objects in the same application, but they have to
    // listen to different ports.
    var apiObj = new API2Go("path/to/config.json");

    // You can also instance a API object without a file, but informing the config
    // object right here, like this:
    var apiObj = new API2Go({
      "DEBUG_MODE"               : 1,
      "NODEJS_LISTEN_PORT"       : 3000,
      "API_FUNCTIONS_MAP"        : "_assets/functions-map.json",
      "GENERAL_LOG"              : "_logs/general.log"
    });

    // You have to register the functions that you want to make invokable
    apiObj.registerFunction("test", function(payload, requestKey, callback, req, res){
      console.log(payload);
      var returnValues = {
        "status": "OK",
        "data": data
      };

      callback(returnValues);
    },
    // This is the map for the function, it defines the path which this request can
    // be invoke and the parameters that will be validated.
    {
      "path": ":id",
      "module": "default",
      "method": "get",
      "params": [
        {
          "paramName": "parameter1",
          "type": "string",
          "mandatory": false,
          "validation": {
            "longerThan": "5",
            "smallerThan": "100"
          }
        },
        {
          "paramName": "parameter2",
          "type": "string",
          "mandatory": false,
          "validation": {
            "longerThan": "60",
            "smallerThan": "65"
          }
        }
      ]
    });

    // The API will start listen in the port you specified in the config.
    apiObj.start();

Every request must return a property *"status"* with *"OK"* or *"ERROR"*.

## Changelog

Refer to `changelog.txt` inside the repo.

## Keep in touch

Eduardo Quagliato, eduardo@quagliato.me

## License

This project is copyrighted by **MIT License** (see *license.txt*).


changelog.txt
v0.0.24 -- 2016-10-14

* Empty _logs directory to fix the test
* Dependencies fix

v0.0.25 -- 2016-12-05

* Language settings;
* New mail function template;
* Nodemailer update;

v0.0.26 -- 2017-01-29

* New options on response (define status code, headers and return a file);
* Update on tests;
* Better algorithm on function registering;

v0.0.27 -- 2017-02-07

* Adding 'nodemailer-html-to-text' node module
* Using 'nodemailer-html-to-text' in send template mail function

v0.0.28 -- 2017-03-06

* Bug fix on requests parameter's validation
* Travis-CI

v0.0.29 -- 2017-07-11

* Possibility to now show README.md on GET /


test.js
//test.js

var async = require('async');
var fs = require('fs');
var request = require('request'); 

var API2Go = require('./api.js');
var apiObj = new API2Go({
  "NODEJS_LISTEN_PORT":"8787",
  "SHOW_README": 0
});

async.series([

  // Register everything
  function(cb) {
    // RPC example
    apiObj.registerFunction("test", function(payload, requestKey, callback, req, apiObj){
      callback({"status": "OK"});
    });

    // REST example
    apiObj.registerFunction("test2", function(payload, requestKey, callback, req, apiObj){
      callback({"status": "OK", "ID": req.params.id});
    });

    apiObj.registerFunction('changelog', function(payload, reqKey, callback, req, apiObj){
      var changelogPath = __dirname + '/changelog.txt';
      fs.readFile(changelogPath, {
        encoding: 'utf8',
        flag: 'r'
      }, function(err, data){
        if (err) {
          return callback({
            status: 'ERROR',
            description: 'Changelog not found'
          }, {
            status: 500
          });
        }

        callback(changelogPath, {
          file: 1
        });
      });
    }, {
      path: 'changelog',
      module: 'test',
      method: 'get',
      params: []
    });

    apiObj.registerFunction('error500', function(payload, reqKey, callback, req, apiObj){
      callback({
        status: 'ERROR',
        description: 'Testing error'
      }, {
        status: 500
      });
    }, {
      path: 'error',
      module: 'test',
      method: 'delete',
      params: []
    });

    apiObj.start();
    cb();
  },

  // 1st test
  function(cb){
    var requestOptions = {
      url: 'http://localhost:' + apiObj.config.NODEJS_LISTEN_PORT + '/test',
      method: 'POST',
      json: {
        parameter1: '123456',
        parameter2: '1234567890123456789012345678901234567890123456789012345678901234'
      }
    };

    request(requestOptions, function(err, httpResponse, body){
      if (err) return cb(err);

      console.log(body);
      console.log(body.status);
      console.log(httpResponse.statusCode);
      console.log(httpResponse.headers);

      if (body.status === 'OK' && 
          httpResponse.statusCode === 200 && 
          httpResponse.headers['content-type'].indexOf('application/json') >= 0) {
        return cb();
      } else {
        return cb(new Error('Problem on first test!'));
      }
    })
  },

  // 2st test
  function(cb){
    var id = parseInt(Math.random() * 1000);
    var requestOptions = {
      url: 'http://localhost:' + apiObj.config.NODEJS_LISTEN_PORT + '/default/' + id,
      method: 'GET'
    };

    request(requestOptions, function(err, httpResponse, body){
      if (err) return cb(err);

      try {
        body = JSON.parse(body);
      } catch (e) {
        return cb(new Error('Response body is not a valid JSON on second test.'));
      }

      console.log(id);
      console.log(body);
      console.log(body.status);
      console.log(httpResponse.statusCode);
      console.log(httpResponse.headers);

      if (body.status === 'OK' && 
          parseInt(body.ID) === id && 
          httpResponse.statusCode === 200 && 
          httpResponse.headers['content-type'].indexOf('application/json') >= 0) {
        return cb();
      } else {
        return cb(new Error('Problem on second test!'));
      }
    })
  },

  // 3rd test
  function(cb){
    var requestOptions = {
      url: 'http://localhost:' + apiObj.config.NODEJS_LISTEN_PORT + '/test/changelog',
      method: 'GET'
    };

    request(requestOptions, function(err, httpResponse, body){
      if (err) return cb(err);

      console.log(body);
      console.log(body.status);
      console.log(httpResponse.statusCode);
      console.log(httpResponse.headers);

      if (httpResponse.statusCode !== 200) return cb(new Error('Third test did not return status 200.'));
      if (httpResponse.headers['content-type'].indexOf('text/plain') === -1) return cb(new Error('Third test did not return text/plain.'));
      if (body.length === 0) return cb(new Error('Third test returned a empty body'));

      cb();
    });
  },

  // 4th test
  function(cb){
    var requestOptions = {
      url: 'http://localhost:' + apiObj.config.NODEJS_LISTEN_PORT + '/test/error',
      method: 'DELETE'
    };

    request(requestOptions, function(err, httpResponse, body){
      console.log(err);
      console.log(httpResponse.statusCode);
      console.log(body);

      if (err) return cb(new Error('Error on fourth test\'s request.'));
      if (httpResponse.statusCode !== 500) return cb(new Error('The fourth test did not return status 500.'));
      try {
        body = JSON.parse(body);
      } catch (e) {
        return cb(new Error('Fourth test did not return a valid JSON.'));
      }
      
      cb();
    });
  },

  // 5th test
  function (cb) {
    var requestOptions = {
      url: 'http://localhost:' + apiObj.config.NODEJS_LISTEN_PORT + '/',
      method: 'GET'
    };

    request(requestOptions, function (err, httpResponse, body) {
      console.log(err);
      console.log(httpResponse.statusCode);
      console.log(body);
      if (err) return cb(new Error('Error on 5th test\'s request'));
      if (httpResponse.statusCode !== 200) return cb(new Error('The 5th test did not return status 200.'));
      if (body.indexOf('It works!') === -1) return cb(new Error('The 5th test did not return the expected body.'));

      cb();
    });
  }
], function(err){
  if (err) {
    console.log('TEST FAILED!')
    console.log(err);
    process.exit(1);
  }

  console.log('TEST SUCCEEDED!');
  process.exit(0);
});


logger.js
// Generic logger module
// Created by Eduardo Quagliato <eduardo@quagliato.me>
// Bauru, Brasil
// 2015-12-26

var fs                       = require("fs");
var moment                   = require("moment");

module.exports = function(message, level, filename) {
  var LOG_FILENAME = config.GENERAL_LOG;
  var MAX_FILESIZE = config.MAX_LOG_FILESIZE; // 10MB

  if (filename === undefined) filename = LOG_FILENAME;
  if (level === undefined) level = "INFO";
  var logTimestamp = moment().format("YYYY-MM-DD HH:mm:ss.SSS ZZ");

  if (filename !== undefined) {
    fs.lstat(filename, function(err, stats){
      if (err) {
        print(filename, message, level, logTimestamp);
      } else {  
        if (stats.size >= MAX_FILESIZE) {
          rotate(filename, function(){
            print(filename, message, level, logTimestamp);
          });
        } else {
          print(filename, message, level, logTimestamp);
        }
      }
    });

    function print(filename, message, level, logTimestamp) {
      var formattedMessage = 
          "[" + logTimestamp + "] " + 
          "[" + level.toUpperCase() + "] " + 
          message;
      fs.appendFileSync(filename, formattedMessage + "\n", { encoding: "utf8", "mode": 644 });
      if (level.toUpperCase() == "CRITICAL" || config.DEBUG_MODE == 1) {
        console.log(formattedMessage);
        if (typeof config.DEBUG_LOG !== "undefined") {
          fs.appendFileSync(config.DEBUG_LOG, formattedMessage + "\n", { encoding: "utf8", "mode": 644 });
        }
      }
    }

     function rotate(filename, callback) {
      fs.rename(filename, filename + "." + moment().format("YYYYMMDDHHmmssSSSZZ"), function(err, data){
        if (!err) {
          callback();
        }
      });
    }
  }
}

src/mail.js
'use strict'

// API2GO - Mail
// 2017-03-07, Curitiba - Brazil
// Author: Eduardo Quagliato<eduardo@quagliato.me>

// Dependencies
const fs                     = require('fs');
const htmlToText             = require('nodemailer-html-to-text').htmlToText;
const moment                 = require('moment');
const nodemailer             = require("nodemailer");
const smtpTransport          = require('nodemailer-smtp-transport');

/*
 * 2017-03-07, Curitiba - Brazil
 * Author: Eduardo Quagliato<eduardo@quagliato.me>
 * Description: The default mail function
 */
module.exports = function(apiObj, mailOptions, mailTemplate, context, callback){

    const validateEmail = function(address){
      let ok = true;
      if (address.indexOf("@") <= 0) ok = false;
      let postAt = address.substr(address.indexOf("@"));
      if (postAt.indexOf(".") <= 0) ok = false;

      return ok;
    }

    if (typeof callback !== "function") {
      apiObj.logger("The callback for sendMail must be a function.", "CRITICAL");
      callback(false);
    }

    if (!mailOptions.hasOwnProperty('to') || mailOptions['to'] == "" || 
        !mailOptions.hasOwnProperty('from_name') || mailOptions['form_name'] == "" || 
        !mailOptions.hasOwnProperty('from') || mailOptions['from'] == "" || 
        !mailTemplate.hasOwnProperty('subject') || mailTemplate['subject'] == "" || 
        !mailTemplate.hasOwnProperty('html') || mailTemplate['html'] == "") {
      apiObj.logger("In order to send an email, to address, from address, from name, email subject and HTML content are required.", "CRITICAL");
      callback(false);
    }

    if (!validateEmail(mailOptions['to'])) {
      apiObj.logger("The {0} address is not valid.".format(mailOptions['to']), "CRITICAL");
      callback(false);
    }

    if (!validateEmail(mailOptions['from'])) {
      apiObj.logger("The {0} address is not valid.".format(mailOptions['from']), "CRITICAL");
      callback(false);
    }

    if (mailOptions.hasOwnProperty('cc')) {
      if (!validateEmail(mailOptions['cc'])) {
        apiObj.logger("The {0} address is not valid.".format(mailOptions['cc']), "CRITICAL");
        callback(false);
      }
    }

    if (mailOptions.hasOwnProperty('bcc')) {
      if (!validateEmail(mailOptions['bcc'])) {
        apiObj.logger("The {0} address is not valid.".format(mailOptions['bcc']), "CRITICAL");
        callback(false);
      }
    }

    let transporter = nodemailer.createTransport(smtpTransport({
      host: apiObj.config.MAIL_HOST,
      port: apiObj.config.MAIL_PORT,
      secure: apiObj.config.MAIL_SECURE,
      ignoreTLS: apiObj.config.MAIL_IGNORE_TLS,
      auth: {
        user: apiObj.config.MAIL_USER,
        pass: apiObj.config.MAIL_PASSWORD
      }
    }));

    transporter.use('compile', htmlToText());

    let send = transporter.templateSender(mailTemplate);

    mailOptions['replyTo'] = "{0} <{1}>".format(mailOptions['from_name'], mailOptions['from'])
    mailOptions['from'] = "{0} <{1}>".format(mailOptions['from_name'], apiObj.config.MAIL_DEFAULT_FROM_USER);

    send(mailOptions, context, function(error, info){
      if(error){
        apiObj.logger("The '{0}' email to {1} couldn't be sent. Stacktrace: {2}".format(mailTemplate.subject, mailOptions.to, error), "CRITICAL");
      } else {
        apiObj.logger("The '{0}' email to {1} was succefully sent.".format(mailTemplate.subject, mailOptions.to), "INFO");
      }

      callback((error ? true : false));
    });
  };

src/audit.js
'use strict'

// API2GO - Audit
// 2017-03-07, Curitiba - Brazil
// Author: Eduardo Quagliato<eduardo@quagliato.me>

// Dependencies
const moment                 = require('moment');
const sha1                   = require('sha1');

/*
 * 2017-03-07, Curitiba - Brazil
 * Author: Eduardo Quagliato<eduardo@quagliato.me>
 * Description: constructor
 */
function Audit (apiObj) {
  this.apiObj = apiObj;
  this.records = {};
};

/*
 * 2017-03-07, Curitiba - Brazil
 * Author: Eduardo Quagliato<eduardo@quagliato.me>
 * Description: It starts the process of auditoring.
 */
Audit.prototype.start = function (functionName, values) {
  const audit = this;
  const apiObj = this.apiObj;
  const requestInfo = {
    function: functionName,
    values: values,
    begin: moment().format("YYYYMMDDHHmmssSSSZZ")
  };

  var requestKey = sha1(JSON.stringify(requestInfo));
  requestInfo.requestKey = requestKey;

  audit.records[requestKey] = requestInfo;

  apiObj.logger(JSON.stringify(audit.records[requestKey]), 'REQUEST_BEGIN', 'audit');

  return requestKey;
};

/*
 * 2017-03-07, Curitiba - Brazil
 * Author: Eduardo Quagliato<eduardo@quagliato.me>
 * Description: It finishes the process of auditoring.
 */
Audit.prototype.finish = function (requestKey, returnValues, extra) {
  const apiObj = this.apiObj;
  const audit = this;

  try {
    returnValues = JSON.parse(returnValues);
  } catch (e) {
  }

  const requestInfo = audit.records[requestKey];

  requestInfo.end = moment().format("YYYYMMDDHHmmssSSSZZ");

  let beginMoment = moment(requestInfo.begin, "YYYYMMDDHHmmssSSSZZ");
  let endMoment = moment(requestInfo.end, "YYYYMMDDHHmmssSSSZZ");

  const minutes = endMoment.diff(beginMoment, "minutes");
  const seconds = endMoment.diff(beginMoment, "seconds");
  const milliseconds = endMoment.diff(beginMoment, "milliseconds");

  requestInfo.duration = `${minutes}m${(seconds - (minutes * 60))}s${(milliseconds - (seconds * 1000))}ms`;
  requestInfo.returnValues = returnValues;
  requestInfo.extra = extra;

  var auditInfo = {};
  auditInfo[requestKey] = requestInfo;

  apiObj.logger(JSON.stringify(auditInfo), "REQUEST-END", 'audit');
};
  

module.exports = Audit;

src/config_parsing.js
'use strict'

// API2GO - Configuration parsing
// 2017-03-07, Curitiba - Brazil
// Author: Eduardo Quagliato<eduardo@quagliato.me>

// Dependencies
const fs                     = require('fs');

module.exports = {
  /*
   * 2017-03-07, Curitiba - Brazil
   * Author: Eduardo Quagliato<eduardo@quagliato.me>
   * Description: Reads the configuration from a file.
   */
  readFromFile: function (configFilepath) {
    var fileConfigs = {};

    var fileData = fs.lstatSync(configFilepath);
    if (!fileData) {
        console.log(`Couldn't find ${configFilepath}.`);
    } else {
      console.log(`Loading config file ${configFilepath}`);
      var fileContent = fs.readFileSync(configFilepath);
      if (!fileContent) {
        console.log(`Couldn't find config file.`);
      } else {
        var configJSON = JSON.parse(fileContent);
        for (var key in configJSON) {
          fileConfigs[key] = configJSON[key];
        }
      }
    }

    return fileConfigs;
  },

  /*
   * 2017-03-07, Curitiba - Brazil
   * Author: Eduardo Quagliato<eduardo@quagliato.me>
   * Description: Merges two sets of configuration
   */
  mergeConfig: function (primary, secondary) {
    var configs = primary;
    for (var key in configs) {
      if (secondary !== undefined && secondary.hasOwnProperty(key)) {
        configs[key] = secondary[key];
      }
    }

    if (secondary !== undefined) {
      for (var key in secondary) {
        if (!configs.hasOwnProperty(key)) {
          configs[key] = secondary[key];
        }
      }
    }

    return configs;
  },

  /*
   * 2017-03-07, Curitiba - Brazil
   * Author: Eduardo Quagliato<eduardo@quagliato.me>
   * Description: Loads the default configuration and the custom and merges it.
   */
  loadConfigFile: function (preseted, loadCallback) {
    let configs = {};

    const defaultConfigs = this.readFromFile(`${__dirname}/../_assets/config-default.json`);
    if (typeof preseted === 'string') {
      const customConfigs = this.readFromFile(preseted);
      configs = this.mergeConfig(defaultConfigs, customConfigs);
    } else {
      if (preseted !== undefined) configs = preseted;
      configs = this.mergeConfig(defaultConfigs, preseted);
    }

    return configs;
  }
}

src/default_logger.js
'use strict'

// API2GO - Default logger
// 2017-03-07, Curitiba - Brazil
// Author: Eduardo Quagliato<eduardo@quagliato.me>

// Dependencies
const fs                     = require('fs');
const moment                 = require('moment');

/*
 * 2017-03-07, Curitiba - Brazil
 * Author: Eduardo Quagliato<eduardo@quagliato.me>
 * Description: The default log engine, using local filesystem.
 */
module.exports = function (apiObj, message, level, bucket) {
  let filename = apiObj.config.LOG.buckets['all'] || 'general.log';
  if (bucket !== undefined) {
    if (!apiObj.config.LOG.buckets.hasOwnProperty(bucket.toLowerCase())) {
      console.log(`The bucket ${bucket} is not set up in the configuration file.`);
      return false;
    }

    filename = apiObj.config.LOG.buckets[bucket.toLowerCase()];
  }

  const maxFilesize = apiObj.config.LOG.max_filesize || 10485760; // 10MB
  const logTimestamp = moment().format('YYYY-MM-DD HH:mm:ss.SSS ZZ');

  if (level === undefined) level = 'INFO';

  fs.lstat(filename, function (err, stats){
    if (!err && stats.size >= maxFilesize) {
      return rotate(filename, function(err){
        if (err) {
          throw err;
        }

        print(filename, message, level, logTimestamp);
      });
    }
    print(filename, message, level, logTimestamp);
  });

  /*
    * 2017-03-07, Curitiba - Brazil
    * Author: Eduardo Quagliato<eduardo@quagliato.me>
    * Description: Actual writing on the file
    */
  function print (filename, message, level, logTimestamp) {
    let onlyFilename = filename;
    while (onlyFilename.indexOf('/') >= 0) onlyFilename = onlyFilename.substr(onlyFilename.indexOf('/') + 1);

    mkdirp(`${__dirname}/../${filename}`.replace(onlyFilename, ''));

    const formattedMessage = `[${logTimestamp}] [${level.toUpperCase()}] ${message}`;
    fs.appendFileSync(filename, formattedMessage + '\n', { encoding: "utf8", "mode": 644 });
    if (level.toUpperCase() == "CRITICAL" || apiObj.config.DEBUG_MODE == 1) {
      console.log(formattedMessage);
    }
  }

  /*
    * 2017-03-07, Curitiba - Brazil
    * Author: Eduardo Quagliato<eduardo@quagliato.me>
    * Description: If the log file hits the max_filesize, it rotates it.
    */
  function rotate (filename, callback) {
    fs.rename(filename, filename + "." + moment().format("YYYYMMDDHHmmssSSSZZ"), function(err, data){
      if (err) return callback(err);
      callback();
    });
  }

  /*
    * 2017-03-07, Curitiba - Brazil
    * Author: Eduardo Quagliato<eduardo@quagliato.me>
    * Description: It creates directores recursively.
    */
  function mkdirp (destination) {
    const directories = destination.split('/');
    let composedPath = '';
    if (directories.length > 0) {
      for (let i = 0; i < directories.length; i++) {
        if (directories[i].trim() === '') continue;

        composedPath += `/${directories[i]}`;
        if (!fs.existsSync(composedPath)) {
          fs.mkdirSync(composedPath);
        }
      }
    }

    return true;
  }
};

src/request_validation.js
'use strict'

// API2GO - Request validation
// 2017-03-07, Curitiba - Brazil
// Author: Eduardo Quagliato<eduardo@quagliato.me>

// Dependencies
const moment                 = require('moment');
const sha1                   = require('sha1');

/*
 * 2017-03-07, Curitiba - Brazil
 * Author: Eduardo Quagliato<eduardo@quagliato.me>
 * Description: validates request's body based on its mapping.
 */
module.exports = function(apiObj, functionName, requestBody, callback) {
  var mapPath = apiObj.config.API_FUNCTIONS_MAP;
  var functionsMap = apiObj.functionsMap;
  if (functionsMap.hasOwnProperty(functionName)) {
    var validationErrors = [];
    const functionSpecs = functionsMap[functionName];

    // Iterate all the parameters specified in the maps.
    for (var param in functionSpecs["params"]) {
      param = functionSpecs["params"][param];

      // The parameter is mandatory but is not in the request content?
      if (!requestBody.hasOwnProperty(param["paramName"]) &&
          param["mandatory"] == true) {
        // Error - Mandatory parameter not present in the request
        validationErrors[validationErrors.length] = {
          "param": param["paramName"],
          "code": "VAL0001",
          "description": apiObj._lang.validation.mandatory_parameter_not_found
        };

      } else {
        var paramValue = requestBody[param["paramName"]];

        if (paramValue !== undefined && paramValue !== null){
          // Validation for string
          if (param["type"] == "string") {
            paramValue = paramValue.trim();

            if (param.hasOwnProperty(["validation"]) &&
                param["validation"].hasOwnProperty("longerThan") &&
                paramValue.length < parseInt(param["validation"]["longerThan"])) {
              // Error - String length smaller then needed
              validationErrors[validationErrors.length] = {
                "param": param["paramName"],
                "code": "VAL1001",
                "description": apiObj._lang.validation.string_length_smaller
              };
            }
            if (param.hasOwnProperty(["validation"]) &&
                param["validation"].hasOwnProperty("longerThan") &&
                paramValue.length > parseInt(param["validation"]["smallerThan"])) {
              // Error - String length longer then needed
              validationErrors[validationErrors.length] = {
                "param": param["paramName"],
                "code": "VAL1002", 
                "description": apiObj._lang.validation.string_length_larger
              };
            }

          // validation for integer
          } else if (param["type"] == "int") {
            try {
              paramValue = parseInt(paramValue);

              if (param.hasOwnProperty(["validation"]) &&
                  param["validation"].hasOwnProperty("greaterThan") &&
                  paramValue < parseInt(param["validation"]["greaterThan"])) {
                // Error - String length smaller then needed
                validationErrors[validationErrors.length] = {
                  "param": param["paramName"],
                  "code": "VAL2001",
                  "description": apiObj._lang.validation.integer_too_small
                };
              }
              if (param.hasOwnProperty(["validation"]) &&
                  param["validation"].hasOwnProperty("lesserThan") &&
                  paramValue > parseInt(param["validation"]["lesserThan"])) {
                // Error - String length longer then needed
                validationErrors[validationErrors.length] = {
                  "param": param["paramName"],
                  "code": "VAL2002",
                  "description": apiObj._lang.validation.integer_too_big
                };
              }

            } catch (e) {
              validationErrors[validationErrors.length] = {
                "param": param["paramName"],
                "code": "VAL0002",
                "description": apiObj._lang.validation.value_not_integer
              };
            }
          }
        }
      }
    }

    return (validationErrors.length > 0 ? validationErrors : undefined);
  } else {
    // Reads the functions-map
    var fileContent = fs.readFileSync("_assets/functions-map.json");
    var functionsMap = JSON.parse(fileContent);
    var validationErrors = [];

    // Verify if the requested function exists in the map.
    if (!functionsMap.hasOwnProperty(functionName)) {
      // Error - Function not found
      validationErrors[validationErrors.length] = {"code": "VAL0000", "description": apiObj._lang.validation.function_not_found};
    } else {
      const functionSpecs = functionsMap[functionName];

      // Iterate all the parameters specified in the maps.
      for (var param in functionSpecs["params"]) {
        param = functionSpecs["params"][param];

        // The parameter is mandatory but is not in the request content?
        if (!requestBody.hasOwnProperty(param["paramName"]) &&
            param["mandatory"] == true) {
          // Error - Mandatory parameter not present in the request
          validationErrors[validationErrors.length] = {
            "param": param["paramName"],
            "code": "VAL0001",
            "description": apiObj._lang.validation.mandatory_parameter_not_found
          };

        } else if (requestBody.hasOwnProperty(param["paramName"])) {
          var paramValue = requestBody[param["paramName"]];

          // Validation for string
          if (param["type"] == "string") {
            paramValue = paramValue.trim();

            if (param.hasOwnProperty(["validation"]) &&
                param["validation"].hasOwnProperty("longerThan") &&
                paramValue.length < parseInt(param["validation"]["longerThan"])) {
              // Error - String length smaller then needed
              validationErrors[validationErrors.length] = {
                "param": param["paramName"],
                "code": "VAL1001",
                "description": apiObj._lang.validation.string_length_smaller
              };
            }
            if (param.hasOwnProperty(["validation"]) &&
                param["validation"].hasOwnProperty("longerThan") &&
                paramValue.length > parseInt(param["validation"]["smallerThan"])) {
              // Error - String length longer then needed
              validationErrors[validationErrors.length] = {
                "param": param["paramName"],
                "code": "VAL1002", "description": apiObj._lang.validation.string_length_larger
              };
            }

          // validation for integer
          } else if (param["type"] == "int") {
            try {
              paramValue = parseInt(paramValue);

              if (param.hasOwnProperty(["validation"]) &&
                  param["validation"].hasOwnProperty("greaterThan") &&
                  paramValue < parseInt(param["validation"]["greaterThan"])) {
                // Error - String length smaller then needed
                validationErrors[validationErrors.length] = {
                  "param": param["paramName"],
                  "code": "VAL2001",
                  "description": apiObj._lang.validation.integer_too_small
                };
              }
              if (param.hasOwnProperty(["validation"]) &&
                  param["validation"].hasOwnProperty("lesserThan") &&
                  paramValue > parseInt(param["validation"]["lesserThan"])) {
                // Error - String length longer then needed
                validationErrors[validationErrors.length] = {
                  "param": param["paramName"],
                  "code": "VAL2002",
                  "description": apiObj._lang.validation.integer_too_big
                };
              }

            } catch (e) {
              validationErrors[validationErrors.length] = {
                "param": param["paramName"],
                "code": "VAL0002",
                "description": apiObj._lang.validation.value_not_integer
              };
            }
          }
        }
      }
    }

    return (validationErrors.length > 0 ? validationErrors : undefined);
  }
};